subsys/console/tty.c

```diff
@@ -149,12 +160,59 @@ static int tty_getchar(struct tty_serial *tty)
 	return c;
 }
 
+static ssize_t tty_read_unbuf(struct tty_serial *tty, void *buf, size_t size)
+{
+	u8_t *p = buf;
+	size_t out_size = 0;
+	int res = 0;
+	u32_t timeout = tty->rx_timeout;
+
+	while (size) {
+		u8_t c;
+		res = uart_poll_in(tty->uart_dev, &c);
+		if (res <= -2) {
+			/* Error occured, best we can do is to return
+			 * accumulated data w/o error, or return error
+			 * directly if none.
+			 */
+			if (out_size == 0) {
+				errno = res;
+				return -1;
+			}
+			break;
+		}
+
+		if (res == 0) {
+			*p++ = c;
+			out_size++;
+			size--;
+		}
+
+		if (size == 0 || (timeout != K_FOREVER && timeout-- == 0)) {
```

Comment:
This conditions is strange. It is obvious that if `timeout == 0` it is `!= K_FOREVER`.
Didn't you mean:
`if (size == 0 || timeout == K_FOREVER || timeout-- == 0)`

Reply:
Well, quite classy C condition ;-). The purpose of `timeout != K_FOREVER &&` is to short-circuit evaluation of `timeout-- == 0` which has a side effect in it.